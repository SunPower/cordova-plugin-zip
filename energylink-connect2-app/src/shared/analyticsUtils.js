import {
  __,
  compose,
  curry,
  gt,
  isEmpty,
  join,
  lensIndex,
  map,
  not,
  over,
  props,
  propSatisfies,
  toLower,
  toUpper,
  when
} from 'ramda'

export const isNotEmpty = compose(not, isEmpty)

export const capitalizeWord = when(
  propSatisfies(gt(__, 0), 'length'),
  compose(join(''), over(lensIndex(0), toUpper), toLower)
)

const stringValueOrEmptyString = stringValue => stringValue || ''

const getStringProps = curry(compose(map(stringValueOrEmptyString), props))

export const getUserProfile = getStringProps([
  'uniqueId',
  'firstName',
  'lastName',
  'email',
  'dealerName',
  'recordType'
])
