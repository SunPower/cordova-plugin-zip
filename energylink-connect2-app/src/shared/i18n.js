import { useMemo } from 'react'
import { useSelector } from 'react-redux'

import * as languages from './locales'
import * as localizedComponents from './localizedComponents'

const getTranslate = translations => (key, ...params) => {
  let value = translations[key] || key
  if (params) {
    params.forEach((p, idx) => {
      value = value.replace(`{${idx}}`, p)
    })
  }
  return value
}

export const useI18n = () => {
  const language = useSelector(state => state.language)
  const { locale = 'en' } = language
  return useMemo(() => getTranslate(languages[locale]), [locale])
}

export const useI18nComponent = () => {
  const language = useSelector(state => state.language)
  const { locale = 'en' } = language || {}

  return useMemo(
    () => componentName => {
      const path = `${locale}_${componentName}`
      const Component = localizedComponents[path] || {}
      if (typeof Component !== 'function') {
        throw new Error(`Component not found: ${path}`)
      }
      return Component()
    },
    [locale]
  )
}
