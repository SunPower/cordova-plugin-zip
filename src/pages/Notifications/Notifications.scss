@import "../../styles/variables";

section.notifications {
  letter-spacing: $letter-spacing-big;
  padding-bottom: 1em !important;
  justify-content: initial;

  .label {
    color: $gray-text-alt2;
    text-transform: uppercase;
    font-size: 0.75rem;
    font-weight: $regular-text;
  }

  .no-notifications {
    font-size: 0.75rem;
    color: $gray-text-alt;
    line-height: 1rem;

    h2,
    .dash {
      text-transform: uppercase;
      font-weight: $bold-text;
    }
  }
}
