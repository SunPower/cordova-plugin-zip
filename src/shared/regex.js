export const PASSWORD_REGEXP = /^(?=.*[a-z])(?=.*\d)[A-Za-z\d]{8,}$/
export const EMAIL_REGEXP = /^\S+@\S+\.\S+$/
export const PHONE_REGEXP = /^(1\s?)?((([0-9]{3}))|[0-9]{3})[\s-]?[0-9]{3}[\s-]?[0-9]{4}$/
