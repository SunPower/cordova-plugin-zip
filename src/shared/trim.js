export const trimObject = obj => {
  return Object.entries(obj)
    .map(([k, v]) => ({
      [k]: typeof v === 'string' || v instanceof String ? v.trim() : v
    }))
    .reduce((acc, kv) => ({ ...acc, ...kv }), {})
}
