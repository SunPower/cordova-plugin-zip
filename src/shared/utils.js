export const either = (condition, whenTrue, whenFalse) =>
  condition ? whenTrue : whenFalse
