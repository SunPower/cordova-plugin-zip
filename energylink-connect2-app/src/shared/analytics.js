import appVersion from '../macros/appVersion.macro'
import { capitalizeWord, getUserProfile } from 'shared/analyticsUtils'
import { MIXPANEL_EVENT_QUEUED } from 'state/actions/analytics'
import { getAppleDeviceFamily } from 'shared/appleDevicesTable'
import { prop } from 'ramda'

const getAppAndDeviceProperties = () => {
  const { device } = window
  const { model, version, platform, manufacturer } = device

  let deviceModel
  let deviceBrowserVersion

  if (platform === 'iOS') {
    deviceModel = getAppleDeviceFamily(model)
    deviceBrowserVersion = version
  } else {
    deviceModel = `${manufacturer}, ${model}`
  }

  return {
    'App Build': appVersion(),
    'App Flavor': process.env.REACT_APP_FLAVOR,
    'OS Version': version,
    ...(deviceBrowserVersion ? { $browser_version: deviceBrowserVersion } : {}),
    $device: deviceModel
  }
}

export const loggedIn = user => {
  const { mixpanel } = window
  const {
    uniqueId: userId,
    firstName,
    lastName,
    email,
    dealerName,
    recordType: dealerType
  } = getUserProfile(user)
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Login', { Success: true })

  return MIXPANEL_EVENT_QUEUED('Login - Success')
}

export const loginFailed = () => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.track('Login', { Success: false })

  return MIXPANEL_EVENT_QUEUED('Login - Failed')
}

export const scanPVS = ({ pvsSN, entryMethod }, event) => {
  const { mixpanel } = window

  mixpanel.register({ 'PVS SN': pvsSN })
  mixpanel.track('Scan PVS Tag', {
    'Entry Method': entryMethod
  })

  return MIXPANEL_EVENT_QUEUED(event)
}

export const siteNotFound = ({
  uniqueId: userId,
  firstName,
  lastName,
  email,
  dealerName,
  recordType: dealerType
}) => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Find Site', { Found: false })

  return MIXPANEL_EVENT_QUEUED('Find Site - site not found')
}

export const siteFound = (
  {
    uniqueId: userId,
    firstName,
    lastName,
    email,
    dealerName,
    recordType: dealerType
  },
  { city, st_id, postalCode, siteKey, commissioned }
) => {
  const { mixpanel } = window
  const appAndDeviceProperties = getAppAndDeviceProperties()

  mixpanel.register(appAndDeviceProperties)
  mixpanel.identify(userId)
  mixpanel.people.set(appAndDeviceProperties)
  mixpanel.people.set({
    $first_name: firstName,
    $last_name: lastName,
    $email: email,
    'Dealer Name': dealerName,
    'Dealer Type': capitalizeWord(dealerType)
  })
  mixpanel.track('Find Site', {
    Found: true,
    ...{
      $city: city,
      State: st_id,
      'Zip Code': postalCode,
      'Site ID': siteKey,
      Commissioned: commissioned
    }
  })

  return MIXPANEL_EVENT_QUEUED('Find Site - site found')
}

export const saveConfiguration = config => {
  const { mixpanel } = window
  mixpanel.track('Configure', config)
  return MIXPANEL_EVENT_QUEUED('Configure - submit site config')
}

export const saveInventory = (inventory, hasEss) => {
  const { mixpanel } = window

  mixpanel.register({
    'Equinox Storage Site': hasEss
  })

  mixpanel.track('Inventory', {
    'AC Module Count': prop('AC_MODULES', inventory),
    'ESS System Size': hasEss ? prop('ESS', inventory) : 'None',
    'Equinox Storage Site': hasEss
  })
  return MIXPANEL_EVENT_QUEUED('Inventory')
}

export const deviceResume = pageResume => {
  const { mixpanel } = window
  mixpanel.register({
    Page: pageResume
  })
  return MIXPANEL_EVENT_QUEUED('User opens app - Register page resume')
}

/**
 *
 * @param {string}[interfaceType] - The type of connection the user has. either can be  PVS, Cell, or Customer WiFi
 * @return {Action<null, null>}
 */
export const internetConnection = (interfaceType = '') => {
  const { mixpanel } = window
  mixpanel.register({
    'Internet Connection': interfaceType
  })
  return MIXPANEL_EVENT_QUEUED('User opens app - Register page resume')
}

export const firmwareUpdateEvent = data => {
  const { mixpanel } = window
  mixpanel.track('Firmware Update', {
    'Current FW Version': prop('fromFWVersion', data),
    'Current BN Version': prop('fromBuildNumber', data),
    'New FW Version': prop('toFWVersion', data),
    'New BN Version': prop('toBuildNumber', data),
    Success: prop('success', data)
  })
  return MIXPANEL_EVENT_QUEUED('Firmware update - Update successful')
}

export const timeMixPanelEvent = name => {
  if (!name) throw new Error('The name parameter is required')
  const { mixpanel } = window
  mixpanel.time_event(name)
}

export const commissionSite = ({ duration, timeConfiguring }) => {
  const { mixpanel } = window
  mixpanel.track('Commission Site', {
    $duration: duration,
    'Time Configuring': timeConfiguring
  })
  return MIXPANEL_EVENT_QUEUED(
    'Commission Site - first successful site configuration'
  )
}
