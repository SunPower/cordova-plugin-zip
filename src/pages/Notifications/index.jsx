import React from 'react'
import moment from 'moment'
import { useSelector, useDispatch } from 'react-redux'
import ModalLayout from '../../components/ModalLayout'
import { useI18n } from '../../shared/i18n'
import ModalNotification from '../../components/ModalNotification'
import { ALERTS_SEEN } from '../../state/actions/alerts'
import { CloudRain } from '../../components/Icons'

import './Notifications.scss'

function renderAlerts(t, alerts) {
  return alerts
    .sort((a, b) => {
      return moment(a.AlertStartTimeUTC).isAfter(moment(b.AlertStartTimeUTC))
        ? -1
        : 1
    })
    .map(alert => {
      const title = t('ALERT_TITLE_TYPE_' + alert.AlertTypeID)
      const description = t('ALERT_TITLE_DESCRIPTION_' + alert.AlertTypeID)
      return (
        <ModalNotification
          key={alert.AlertsID}
          id={alert.AlertsID}
          date={moment(alert.AlertStartTimeUTC)}
          removable={[998, 999].indexOf(alert.AlertsID) === -1} // TODO: remove this after SPI
        >
          <div className="is-flex level file is-centered pl-20 pr-20">
            {[998, 999].indexOf(alert.AlertsID) > -1 ? ( // TODO: remove this after SPI
              <div className="mr-5">
                <CloudRain />
              </div>
            ) : null}
            <span className="has-text-centered is-size-6 has-text-weight-bold has-text-black">
              {title}
            </span>
          </div>
          <div>{description}</div>
        </ModalNotification>
      )
    })
}

function Notifications({ location, history }) {
  const t = useI18n()
  const dispatch = useDispatch()
  const alerts = useSelector(state => state.alerts.data)
  let alertsList = Object.keys(alerts)
    .map(k => alerts[k])
    .filter(alert => !alert.dismissed)

  // TODO: remove this after SPI
  alertsList = alertsList.concat([
    {
      AlertsID: 999,
      AlertTypeID: 999,
      AlertStartTimeUTC: 1569165360000
    },
    {
      AlertsID: 998,
      AlertTypeID: 999,
      AlertStartTimeUTC: 1566450000000,
      seenTimestamp: 1566450000000
    }
  ])

  const unseenAlerts = alertsList.filter(a => !a.seenTimestamp)
  const seenAlerts = alertsList.filter(a => a.seenTimestamp)

  return (
    <ModalLayout
      className="notifications"
      history={history}
      title={t('NOTIFICATIONS')}
      from={location && location.state && location.state.from}
      onClose={() => dispatch(ALERTS_SEEN())}
    >
      {unseenAlerts.length > 0 && (
        <div className="label">{t('MOST_RECENT')}</div>
      )}
      {renderAlerts(t, unseenAlerts)}
      {seenAlerts.length > 0 && <div className="label">{t('OLDER')}</div>}
      {renderAlerts(t, seenAlerts)}
      {unseenAlerts.length === 0 && seenAlerts.length === 0 && (
        <div className="no-notifications has-text-centered mt-100">
          <h2>{t('NO_NOTIFICATIONS_TITLE')}</h2>
          <div className="dash">-</div>
          <p className="pl-20 pr-20">{t('NO_NOTIFICATIONS_TEXT')}</p>
        </div>
      )}
    </ModalLayout>
  )
}

export default Notifications
