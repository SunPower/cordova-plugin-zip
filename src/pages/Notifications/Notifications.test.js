import React from 'react'
import { shallow } from 'enzyme'
import * as reactRedux from 'react-redux'
import * as i18n from '../../shared/i18n'
import Notifications from '.'

describe('Notifications page', () => {
  const alerts = {
    86772640: {
      AlertsID: 86772640,
      AlertTypeID: 2,
      AddressID: 11781,
      AlertStartTimeUTC: '2019-05-04T16:55:00.000Z'
    },
    86772642: {
      AlertsID: 86772642,
      AlertTypeID: 1,
      AddressID: 11781,
      AlertStartTimeUTC: '2017-05-04T16:55:00.000Z'
    },
    86772644: {
      AlertsID: 86772644,
      AlertTypeID: 4,
      AddressID: 11781,
      AlertStartTimeUTC: '2018-05-04T16:55:00.000Z'
    }
  }
  let dispatchMock

  beforeEach(() => {
    dispatchMock = jest.fn()

    jest.spyOn(reactRedux, 'useDispatch').mockImplementation(() => dispatchMock)
    jest.spyOn(reactRedux, 'useSelector').mockImplementation(() => alerts)
    jest
      .spyOn(i18n, 'useI18n')
      .mockImplementation(path => (key, ...params) =>
        `${key.toUpperCase()} ${params.join('_')}`.trim()
      )
  })

  it('renders without crashing', () => {
    const component = shallow(<Notifications />)
    expect(component).toMatchSnapshot()
  })
})
