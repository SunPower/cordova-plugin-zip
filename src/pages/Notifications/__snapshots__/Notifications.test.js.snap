// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications page renders without crashing 1`] = `
<ModalLayout
  className="notifications"
  onClose={[Function]}
  title="NOTIFICATIONS"
>
  <div
    className="label"
  >
    MOST_RECENT
  </div>
  <ModalNotification
    date={"2019-09-22T15:16:00.000Z"}
    id={999}
    key="999"
    removable={false}
  >
    <div
      className="is-flex level file is-centered pl-20 pr-20"
    >
      <div
        className="mr-5"
      >
        <ForwardRef(SvgRainCloud) />
      </div>
      <span
        className="has-text-centered is-size-6 has-text-weight-bold has-text-black"
      >
        ALERT_TITLE_TYPE_999
      </span>
    </div>
    <div>
      ALERT_TITLE_DESCRIPTION_999
    </div>
  </ModalNotification>
  <ModalNotification
    date={"2019-05-04T16:55:00.000Z"}
    id={86772640}
    key="86772640"
    removable={true}
  >
    <div
      className="is-flex level file is-centered pl-20 pr-20"
    >
      <span
        className="has-text-centered is-size-6 has-text-weight-bold has-text-black"
      >
        ALERT_TITLE_TYPE_2
      </span>
    </div>
    <div>
      ALERT_TITLE_DESCRIPTION_2
    </div>
  </ModalNotification>
  <ModalNotification
    date={"2018-05-04T16:55:00.000Z"}
    id={86772644}
    key="86772644"
    removable={true}
  >
    <div
      className="is-flex level file is-centered pl-20 pr-20"
    >
      <span
        className="has-text-centered is-size-6 has-text-weight-bold has-text-black"
      >
        ALERT_TITLE_TYPE_4
      </span>
    </div>
    <div>
      ALERT_TITLE_DESCRIPTION_4
    </div>
  </ModalNotification>
  <ModalNotification
    date={"2017-05-04T16:55:00.000Z"}
    id={86772642}
    key="86772642"
    removable={true}
  >
    <div
      className="is-flex level file is-centered pl-20 pr-20"
    >
      <span
        className="has-text-centered is-size-6 has-text-weight-bold has-text-black"
      >
        ALERT_TITLE_TYPE_1
      </span>
    </div>
    <div>
      ALERT_TITLE_DESCRIPTION_1
    </div>
  </ModalNotification>
  <div
    className="label"
  >
    OLDER
  </div>
  <ModalNotification
    date={"2019-08-22T05:00:00.000Z"}
    id={998}
    key="998"
    removable={false}
  >
    <div
      className="is-flex level file is-centered pl-20 pr-20"
    >
      <div
        className="mr-5"
      >
        <ForwardRef(SvgRainCloud) />
      </div>
      <span
        className="has-text-centered is-size-6 has-text-weight-bold has-text-black"
      >
        ALERT_TITLE_TYPE_999
      </span>
    </div>
    <div>
      ALERT_TITLE_DESCRIPTION_999
    </div>
  </ModalNotification>
</ModalLayout>
`;
